### DISCUSSION TITLE
Best practice when performing realtime data updates

### DISCUSSION THREAD

**Discussion by JamesRexMiller4** [visgl/deck.gl#6274] [Q&A] [answered]
I am working on an application that involves rendering chemical compound concentrations as data points on a map. Working on functionality for realtime, where the dataset accumulates over time, and pre-existing data, where the entire dataset is loaded all at once. 

The all at once iteration is working smooth; I have multiple layers: ColumnLayer, HeatmapLayer, ScatterPlot, and IconLayer, users can toggle between Scatter and Column, at different zoom levels Heatmap's visibility is set to true and other layers to false, Icon layer for windbarbs. (Chef's kiss)

For the realtime data I am running into an issue where I have lets say 15 chemical compounds for a single coordinate position. If I opt for creating a layer for every compound on each socket event, I quickly run into a warning from deck.gl that says only the first 255 layers will be pickable (for rendering a tooltip but also interactivity for onClick handler), and if I opt for creating a layer for each chunk I end up getting "collisions" where all the projections are overlapping each other (lose the offset prop). 

Would venture to guess that either approach I've taken is not ideal, but curious if anyone has recommendations for the most performant way to append data in realtime to an existing dataset? Currently not holding onto the data _outside_ of deck.gl as it comes through a socket stream.

Would it be better to setup layers for all the compounds on the first socket event, and then push the proceeding data to the respective layer so then it remains at +/- 15 (new compounds may appear / existing compounds may disappear from payload)?

### 
Example data:

```
{
   "concentration": 4000, 
   "name": "Acetone",
   "timestamp": new Date(),
   "coordinates" [lng, lat]
 }
 ```

**Comment by chrisgervang** [comment]
What is a "chunk" in the alternative option you tried?

In the "all upfront" case you had working, what is different about the visualization layers you'd like to show? If they're the same set of layer I would suspect pushing data to their respective layer on each socket event is a good way to go.

**Reply by JamesRexMiller4** [reply]
A chunk would be the whole socket event payload, comprised of an array of objects. In the first example I am creating a ColumnLayer for each compound in the payload, and in the latter I create only one ColumnLayer but stick all 15 compounds data in the layer's data prop. 

The former way is how I do it when going the "all upfront" approach, and gives the desired effect of having multiple columns in a region only slightly offset from each other, with the elevation of the column corresponding to the concentration value. 

The main difference between the two approaches I'd say: 

1) The frequency setLayers is called to trigger the redraw. The socket event emits every second, so I am calling setLayers on every emit event vs in the all upfront approach, calling it semi frequently when a user filters which columns/compounds they wish to see.

2) The different types of layers I work with. In the "all upfront" approach I am working with Heatmap, Scatter, Column, and Icon layers, whereas for the realtime approach I am only currently using ColumnLayer. 


For the push to respective layers approach, would there be performance hits? Thought I read in the Performance Optimization section of the Documentation that it's better to append new layers over mutating existing ones because it would trigger unneccessary redrawing of existing data that hadn't changed. Or is this a sliding scale for my use case, since initially the cost of redrawing everything wouldn't be too expensive but would worsen over time?

**Comment by JamesRexMiller4** [comment]
For additional context:
![tooltip-ex](https://user-images.githubusercontent.com/27719824/136288302-44f802dd-6511-4d41-b5d7-acf80a1d768c.png)
![num-layers-before-warning](https://user-images.githubusercontent.com/27719824/136288308-4c46edd7-02ac-4f68-b99f-6fcbc04da048.png)

When taking the "chunk" approach
![chunked-example](https://user-images.githubusercontent.com/27719824/136288571-c857c0c4-5db9-406d-9caf-0146528c07e7.png)

**Comment by mcost45** [comment, marked as answer]
From my experience adding new layers will end up killing performance much, much faster than squashing the data into a single layer - for a decently large dataset.

By default, when you change a layer's data object, the layer will throw all of its calculated attributes out, and recalculate everything from scratch (i.e. positions, colours, etc), because diffing is hard. The time it takes to do that is relative to the size of your data. Adding a new layer avoids these recalcuations - but the more you add, the amount of work required per frame can increase very fast, leading to poor performance. The amount you can visualise by adding a new layer each time is much smaller. There are two ways you can minimise layer count whilst having performant data updates:

The first option you have is to implement [_dataDiff](https://deck.gl/docs/api-reference/core/layer#_datadiff) for your layers. This can help you only recalculate what is necessary on data changes, instead of recalculating from scratch.

The second, and most performant but hardcore option is to manually create your own attributes with TypedArrays and pass these into the layers. Then your layer will take your attributes instead of calculating its own. Working with TypedArrays is faster, but you will have the added inconvencience of managing all the flattened data yourself - i.e. you will have to map all your JSON data into your TypedArrays. From a low-level point of view, TypedArrays are faster in every way (working with tightly-packed memory).

**Reply by JamesRexMiller4** [reply]
Right on, appreciate the insights from you both! Will go forward with this path and if I run into any hiccups you may hear from me again haha
